StyleCop.Analyzers/StyleCop.Analyzers/Helpers/SyntaxTreeHelpers.cs

```diff
@@ -72,29 +75,52 @@ public static bool IsWhitespaceOnly(this SyntaxTree tree, CancellationToken canc
                 && TriviaHelper.IndexOfFirstNonWhitespaceTrivia(firstToken.LeadingTrivia) == -1;
         }
 
-        internal static bool ContainsUsingAlias(this SyntaxTree tree, ConcurrentDictionary<SyntaxTree, bool> cache)
+        internal static bool ContainsUsingAlias(this SyntaxTree tree, IReadOnlyDictionary<SyntaxTree, bool> cache)
         {
             if (tree == null)
             {
                 return false;
             }
 
-            bool result;
-            if (cache.TryGetValue(tree, out result))
+            if (cache == null)
+            {
+                // NOTE: This happens if any syntax tree in the compilation contains a global using alias
+                return true;
+            }
+
+            if (cache.TryGetValue(tree, out var result))
             {
                 return result;
             }
 
-            bool generated = ContainsUsingAliasNoCache(tree);
-            cache.TryAdd(tree, generated);
-            return generated;
+            Debug.Assert(false, "This should not happen. Syntax tree could not be found in cache!");
+            return false;
         }
 
-        private static bool ContainsUsingAliasNoCache(SyntaxTree tree)
+        private static IReadOnlyDictionary<SyntaxTree, bool> CreateDictionary(Compilation compilation)
         {
-            var nodes = tree.GetRoot().DescendantNodes(node => node.IsKind(SyntaxKind.CompilationUnit) || node.IsKind(SyntaxKind.NamespaceDeclaration) || node.IsKind(SyntaxKindEx.FileScopedNamespaceDeclaration));
+            var result = new Dictionary<SyntaxTree, bool>();
 
-            return nodes.OfType<UsingDirectiveSyntax>().Any(x => x.Alias != null);
+            foreach (var tree in compilation.SyntaxTrees)
```

Comment:
Alright. GetImportScopes seems to be supported from Roslyn 4.3 and global usings from 4.0, so I guess that there will be some versions where the bug will still exist if that approach is taken, but some false negatives might be ok since it will only happen for a few versions?

Would this method be supported automatically with newer (than c# 8) light-up support or would it require manually implementing that support?

Is this method quick? I have gotten the impression that the semantic model should be used with care, but maybe this is not a problematic method.